common/src/test/java/org/apache/iceberg/common/TestDynConstructors.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.common;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import org.junit.jupiter.api.Test;
+
+public class TestDynConstructors {
+  @Test
+  public void testImplNewInstance() throws Exception {
+    DynConstructors.Ctor<MyClass> ctor =
+        DynConstructors.builder().impl(MyClass.class).buildChecked();
+    assertThat(ctor.newInstance()).isInstanceOf(MyClass.class);
+  }
+
+  @Test
+  public void testInterfaceImplNewInstance() throws Exception {
+    DynConstructors.Ctor<MyInterface> ctor =
+        DynConstructors.builder(MyInterface.class)
+            .impl("org.apache.iceberg.common.TestDynConstructors$MyClass")
+            .buildChecked();
+    assertThat(ctor.newInstance()).isInstanceOf(MyClass.class);
+  }
+
+  @Test
+  public void testInterfaceWrongImplString() throws Exception {
+    DynConstructors.Ctor<MyInterface> ctor =
+        DynConstructors.builder(MyInterface.class)
+            // TODO this should throw, since the MyUnrelatedClass does not implement MyInterface
```

Comment:
still a TODO left here

Reply:
This is reflection on current state of the API, pre-existing problem.
I can remove this TODO if this helps. I did not plan to address this as part of this cleanup PR.

Reply:
BTW my thinking about TODO in the test is this -- a test should prevent undesired behavioral changes of the implementation. If i write a test exercising specific behavior and I don't think the behavior change should be prevented I put a comment like this one. If I don't put the comment in, someone may treat this test as asserting something important and thus give up an otherwise good change.
I understand this is a bit convoluted, but that's a lesson learned from Trino on how contributors sometimes react to test failures. I've seen test failures causing otherwise good code to become less good, because the contributor considered modifying the test is not an option.

Alternatively, I could not have the test, but I am not convinced this is a better option.

Reply:
@nastra does this explanation address your concerns? or what should i change here?

Reply:
I think leaving the TODO here is fine and we should probably address this in a follow up so that it throws when `buildChecked` is called

Reply:
good point!